support-frontend/app/actions/AsyncAuthenticatedBuilder.scala

```diff
@@ -0,0 +1,41 @@
+package actions
+
+import com.typesafe.scalalogging.LazyLogging
+import play.api.mvc.Security.AuthenticatedRequest
+import play.api.mvc._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+// scalastyle:off
+// This is based on AuthenticationBuilder in https://github.com/playframework/playframework/blob/1ca9d0af237ac45a011671b2e036d726cd05e4e7/core/play/src/main/scala/play/api/mvc/Security.scala
+// The main difference is that this action builder enables authenticating a user asynchronously.
+// i.e. RequestHeader => Future[Option[U]] instead of RequestHeader => Option[U]
+// In the context of Guardian applications, asynchronous authentication equals making a call to identity API,
+// to authenticate the user data.
+// TODO: considering porting this to identity-play-auth.
+// scalastyle:on
+class AsyncAuthenticatedBuilder[U](
+    userinfo: RequestHeader => Future[Option[U]],
+    defaultParser: BodyParser[AnyContent],
+    onUnauthorized: RequestHeader => Result
+)(implicit val executionContext: ExecutionContext)
+  extends ActionBuilder[({ type R[A] = AuthenticatedRequest[A, U] })#R, AnyContent] with LazyLogging { // scalastyle:ignore
+
+  lazy val parser = defaultParser
```

Comment:
it's risky in scala to use lazy vals and LazyLogging together with Futures.  lazy would get a (java) lock on the containing object, which could deadlock if there is a future that is mapped or flatmapped between the two.
(see https://alvinalexander.com/scala/look-how-scala-lazy-val-converted-to-java-code-jvm-bytecode )
In this case it will be fine, as the value of parser is a static value, but if the defaultParser was a code block or a method call, things get more risky.
We have seen this issue in membership-frontend.
There was a SIP to change that But I'm not sure if it got anywhere: https://docs.scala-lang.org/sips/improved-lazy-val-initialization.html#version-6---no-synchronization-on-this-and-concurrent-initialization-of-fields

Reply:
So something like:
```scala
lazy val parser = {
  logger.info("initialising parser")
  defaultParser
}
```
would not be cool?

Reply:
not that simple, the problem would actually be
```scala
lazy val parser = Future.successful(()).map{ _ =>
  logger.info("initialising parser")
  defaultParser
}
```
so basically the lazy val gets the lock, the `map` on the future uses a different thread, and the `logger.` would try to initialise another lazy val and get the same lock

Reply:
ah I see - so because lazy vals lock on the surrounding class instance, the synchronous version wouldn't be a problem (since the thread that has a lock on the class instance is the same thread that's evaluating the logging statement) (?)

Reply:
yes exactly, that would be fine.
I'm not sure why they don't just create one object for each lazy val, maybe just perf reasons.  Even in my java days I used to just create an object for synchronising if I needed to do it rather than relying on something I didn't control.